File: jutari/src/games/RomSettings.jl
"""
    RomSettings

Per-game scoring + termination rules. The `RomSettings` abstract type
is the interface front-ends use to read out reward / lives / game-over
from a running `Console`. Per-game subtypes inspect the console's RAM
or registers at game-specific addresses; that's reverse-engineering
work, deferred beyond P6.

`GenericRomSettings` is a no-op stub: zero reward, never terminal.
"""
module RomSettingsModule

using ..ConsoleModule: Console

export RomSettings, GenericRomSettings,
       romsettings_reset!, romsettings_is_terminal,
       romsettings_get_reward, romsettings_lives,
       romsettings_uses_paddles, romsettings_swap_paddles,
       romsettings_starting_actions, romsettings_difficulty

abstract type RomSettings end

# Required interface — subtypes override these via Julia multiple dispatch.

romsettings_reset!(::RomSettings)             = nothing
romsettings_is_terminal(::RomSettings, ::Console) = false
romsettings_get_reward(::RomSettings, ::Console) = 0
romsettings_lives(::RomSettings, ::Console)      = 0
# Per-game starting actions emulated by xitari's `StellaEnvironment::reset`
# AFTER the 60-NOOP + 4-RESET boot burn and the settings.reset() call
# (only when settings.getBool("use_starting_actions") is true — default
# in xitari is `true`). Pitfall + Enduro use this to put the agent into
# a known initial pose (PLAYER_A_UP / PLAYER_A_FIRE respectively); without
# emulating those frames we diverge from xitari by 1 frame of state.
# Default empty — joystick games without a startup pose return `Int[]`.
# See `xitari/games/supported/Pitfall.cpp::getStartingActions` etc.
romsettings_starting_actions(::RomSettings)      = Int[]
# Default: joystick. Override `romsettings_uses_paddles(::MyType) =
# true` in per-game subtypes whose stella.pro entry has
# Controller.Left/Right "PADDLES" — `StellaEnvironment` reads this
# to auto-translate LEFT/RIGHT actions into INPT0 dump-pot
# paddle-position changes. Mirror of xitari's `m_use_paddles`
# auto-detection from the cart's properties.
romsettings_uses_paddles(::RomSettings)          = false
# Default: paddles not swapped (Breakout convention). Override to
# `true` on per-game subtypes whose stella.pro entry has
# `Controller.SwapPaddles "YES"` (Pong / Video Olympics being the
# headline case). With the swap, xitari's `Paddles` controller wires
# `PaddleZeroResistance` (the user-driven paddle on Pin Five via
# `applyActionPaddles`) to controller Pin Five — which the TIA then
# reads as `INPT1` (left-jack Pin Five → INPT1 per
# `xitari/emucore/TIA.cxx::peek` case 0x09). So for swapped paddle
# games, our `_apply_paddle_action!` must write the user paddle to
# `paddle_resistance[1]` (INPT1) instead of `paddle_resistance[0]`
# (INPT0).
romsettings_swap_paddles(::RomSettings)          = false
# Console difficulty switches (SWCHB bits 0x40 = P0/Left, 0x80 = P1/Right).
# Returns `(p0_difficulty_a, p1_difficulty_a)`. xitari's default properties
# (Props.cxx) set BOTH difficulties to "B" → SWCHB 0x3F, so the default here
# is `(false, false)` (B/B), matching the 58 bit-exact games. Per-game
# subtypes whose stella.pro entry overrides a difficulty to "A" return
# `true` for that switch (e.g. Amidar = A/A → SWCHB 0xFF). xitari
# Switches.cxx: a "B" property clears the bit, otherwise the bit stays set.
romsettings_difficulty(::RomSettings)            = (false, false)

"""No-op RomSettings — never terminal, zero reward, joystick-only."""
mutable struct GenericRomSettings <: RomSettings
    GenericRomSettings() = new()
end

end # module

File: jaxtari/jaxtari/games/rom_settings.py
"""Base `RomSettings` interface for per-game scoring + termination.

Implementations override `is_terminal`, `get_reward`, and (optionally)
`lives`. Each is called by `StellaEnvironment.step` after the frame
finishes; the settings object reads the console's RAM / RIOT / cart
state to compute its return value.

`reset` is invoked at the start of each episode so the settings can
clear any running score / lives counters.

A subclass may need to retain state between frames (e.g. last frame's
score, to compute the per-step reward delta). For that we deliberately
keep RomSettings plain mutable Python — there's no JAX tracing path
through reward computation.
"""

from __future__ import annotations

from typing import Protocol, runtime_checkable

from jaxtari.console import Console


@runtime_checkable
class RomSettings(Protocol):
    """Per-game scoring + termination rules. Plain Python, not a PyTree."""

    def reset(self) -> None:
        """Called at the start of each episode."""

    def is_terminal(self, console: Console) -> bool:
        """Return True when the game is over."""

    def get_reward(self, console: Console) -> int:
        """Return the per-step reward (score delta since the previous call)."""

    def lives(self, console: Console) -> int:
        """Return the player's remaining lives, or 0 if the game has no
        explicit life counter."""

    def uses_paddles(self) -> bool:
        """Return True when the game expects paddle controllers
        (Breakout, Pong/Video Olympics, Warlords, Casino, …) so
        `StellaEnvironment` translates LEFT/RIGHT actions into
        INPT0 dump-pot paddle-position changes. xitari reads this
        from stella.pro's `Controller.Left "PADDLES"`; jaxtari
        encodes it directly on the per-game settings class.
        Default `False` matches the majority of ROMs (joystick-only).
        """

    def swap_paddles(self) -> bool:
        """Return True when the game's stella.pro entry has
        `Controller.SwapPaddles "YES"` (Pong / Video Olympics).
        With SwapPaddles=YES, the user's (left-controller) paddle is
        wired to INPT1 instead of INPT0. xitari handles this inside
        `Paddles::Paddles(jack, event, swap)` via the
        `myPinEvents[2/3][0/1]` wiring table. Default `False`
        matches Breakout / Warlords / Casino — only Pong (Video
        Olympics) flips this. Mirrors jutari's
        `romsettings_swap_paddles`.
        """


class GenericRomSettings:
    """No-op stub: reward = 0, never terminal, no lives, joystick-only.

    Use this when you want to run a ROM without per-game scoring (e.g.
    for collecting raw screen frames or building a goldens dataset).
    """

    def reset(self) -> None:
        return None

    def is_terminal(self, console: Console) -> bool:
        return False

    def get_reward(self, console: Console) -> int:
        return 0

    def lives(self, console: Console) -> int:
        return 0

    def uses_paddles(self) -> bool:
        # Default: assume joystick. Override in per-game subclasses
        # whose stella.pro entry has Controller.Left/Right "PADDLES".
        return False

    def swap_paddles(self) -> bool:
        # Default: assume no swap. Override in per-game subclasses
        # whose stella.pro entry has Controller.SwapPaddles "YES"
        # (currently only Pong / Video Olympics).
        return False
